src/main/java/at/porscheinformatik/sonarqube/licensecheck/npm/PackageJsonDependencyScanner.java

```diff
@@ -46,46 +48,52 @@ public PackageJsonDependencyScanner(Settings settings) {
             File nodeModulesFolder = new File(packageJsonFile.getParentFile(), "node_modules");
             if (!nodeModulesFolder.exists() || !nodeModulesFolder.isDirectory())
             {
-                return Collections.emptyList();
+                return Collections.emptySet();
             }
 
             try (InputStream fis = new FileInputStream(packageJsonFile); JsonReader jsonReader = Json.createReader(fis))
             {
                 Deque<String> transitiveStack = new ArrayDeque<>();
-                return new ArrayList<>(getDependenciesFrom(jsonReader.readObject(), nodeModulesFolder, transitiveStack));
+                return getDependenciesFrom(jsonReader.readObject(), nodeModulesFolder);
             }
             catch (IOException e)
             {
                 LOGGER.error("Error reading package.json", e);
             }
         }
-        return Collections.emptyList();
+        return Collections.emptySet();
     }
 
-    private Set<Dependency> getDependenciesFrom(JsonObject packageJsonObject, File nodeModulesFolder, Deque<String> transitiveStack)
+    private Set<Dependency> getDependenciesFrom(JsonObject packageJsonObject, File nodeModulesFolder)
     {
         JsonObject jsonObjectDependencies = packageJsonObject.getJsonObject("dependencies");
         if (jsonObjectDependencies != null)
         {
-            return dependencyParser(jsonObjectDependencies, nodeModulesFolder, transitiveStack);
+            return dependencyParser(jsonObjectDependencies, nodeModulesFolder);
         }
         return Collections.emptySet();
     }
 
-    private Set<Dependency> dependencyParser(JsonObject jsonDependencies, File nodeModulesFolder, Deque<String> transitiveStack)
+    private Set<Dependency> dependencyParser(JsonObject jsonDependencies, File nodeModulesFolder)
     {
         Set<Dependency> dependencies = new LinkedHashSet<>();
 
         for (String packageName : jsonDependencies.keySet())
         {
-            moduleCheck(nodeModulesFolder, packageName, dependencies, transitiveStack);
+            moduleCheck(nodeModulesFolder, packageName, dependencies);
         }
 
         return dependencies;
     }
 
-    private void moduleCheck(File nodeModulesFolder, String packageName, Set<Dependency> dependencies, Deque<String> transitiveStack)
+    private void moduleCheck(File nodeModulesFolder, String packageName, Set<Dependency> dependencies)
     {
+        if (dependencies.stream().anyMatch(d -> packageName.equals(d.getName())))
+        {
+            LOGGER.warn("Circular dependency detected in {}. Current stack of transitive deps: {}", packageName, dependencies);
```

Comment:
@derkoe I'm afraid this doesn't work.  Each time you enter `dependencyParser()` you get a new `Set<Dependency>`.  So this `dependencies` is only a set of *immediate dependencies at the current level*.

As a result it can't be used to track circular dependencies - which is why I was separately maintaining a stack which is *only initialised once*.

I have just pushed a small commit to prove this in the unit tests - you should see the build of this PR start failing.

I suppose an alternative would be to initialise the set only once - and pass it in to `getDependenciesFrom()` in the same way I did with the `Deque` - but then it would no longer be appropriate to log a warning like this, as repeats in that circumstance could just be multiple artifacts transitively depending on the same thing (which is perfectly normal).

Reply:
Thanks for checking - yes I completely misunderstood this part. Anyway I refactored the scanning now. See below.